### PR TITLE
NineDoF API

### LINKED ISSUE #9: Implement NineDoF API
Implement [NineDoF](https://github.com/tock/tock/blob/master/capsules/extra/src/ninedof.rs) API

### PR DESCRIPTION
**Pull request by SheepSeb**
## Pull Request Overview

This pull request adds:
- a 9-DOF API resolves #9  similar to the [libtock-c version](https://github.com/tock/libtock-c/blob/master/libtock/ninedof.h). The API provides a function that simulates a synchronous read of the accelerometer, magnetometer and gyroscope, also functions to initiate a reading, register/unregister a listener, check for driver existence, calculate the magnitude of the accelerometer
- an example app which samples the 9-DOF sensor every 700ms and prints the measured x,y,z for the accelerometer, magnetometer and gyroscope as well as the magnitude of the accelerometer
- a ```fake``` NineDoF driver, for testing
- unit tests for the fake driver and for the user space library

## Testing Strategy
This PR was tested with the unit tests and by running the example app on a microbit v2 board

## TODO or Help Wanted
Feedback

## Documentation Updated
- [x] no updates are required

### CODE REVIEW

File: apis/ninedof/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+                *z = data.z;
+                Ok(())
+            }
+        }
+    }
+
+    /// Calculate the magnitude of the accelerometer reading
+    /// Returns value of magnitude if the operation was successful
+    /// Returns 0.0 if the operation was unsuccessful
+    pub fn ninedof_read_accel_mag() -> f64 {
```
Comment (alexandruradovici) [inline apis/ninedof/src/lib.rs:159]:
```suggestion
    pub fn read_accelerometer_mag() -> f64 {
```

Comment (alexandruradovici) [reply apis/ninedof/src/lib.rs:159]: Use the `read_accelerometer` function.

Comment (alexandruradovici) [inline apis/ninedof/src/lib.rs:159]: Magnitudes are valid only for the accelerometer?

File: apis/ninedof/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                *z = data.z;
+                Ok(())
+            }
+        }
+    }
+
+    /// Initiate a synchronous magnetometer measurement.
+    /// Returns Ok(data) if the operation was successful
+    /// Returns Err(ErrorCode) if the operation was unsuccessful
+    pub fn read_magnetometer_sync(x: &mut i32, y: &mut i32, z: &mut i32) -> Result<(), ErrorCode> {
```
Comment (alexandruradovici) [inline apis/ninedof/src/lib.rs:101]:
```suggestion
    pub fn read_magnetometer_sync() -> Result<NineDofData, ErrorCode> {
```

File: examples/ninedof.rs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            )
+            .unwrap(),
+            Err(_) => writeln!(Console::writer(), "error while reading magnetometer").unwrap(),
+        }
+
+        let acc_mag = NineDof::ninedof_read_accel_mag();
+        writeln!(Console::writer(), "Magnitude of acceleration: {}", acc_mag).unwrap();
+
+        Alarm::sleep_for(Milliseconds(700)).unwrap();
+    }
```
Comment (alexandruradovici) [inline examples/ninedof.rs:65]: Move this into the `Ok(_)` branch.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Comment by alexandruradovici** [comment]
@SheepSeb any updates?

File: apis/ninedof/src/lib.rs (hunk 1 above)
Comment (SheepSeb) [reply apis/ninedof/src/lib.rs:144]: I haven't seen them implemented in libtock-c for the [9-DOF](https://github.com/tock/libtock-c/blob/master/libtock/ninedof.c#L24), I assume that for the magnetometer and gyroscope the magnitudes do not hold any information that could be useful, but for the accelerometer you could use it to get the total acceleration

**Review by alexandruradovici — CHANGES_REQUESTED** [review]
Please rebase this.

**Review by alexandruradovici — APPROVED** [review]

File: apis/ninedof/src/lib.rs (hunk 1 above)
Comment (alexandruradovici) [reply apis/ninedof/src/lib.rs]: Ok.

File: apis/ninedof/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+                *z = data.z;
+                Ok(())
+            }
+        }
+    }
+
+    /// Initiate a synchronous gyroscope measurement.
+    /// Returns Ok(data) as NineDofData if the operation was successful
+    /// Returns Err(ErrorCode) if the operation was unsuccessful
+    pub fn read_gyro_sync(x: &mut i32, y: &mut i32, z: &mut i32) -> Result<(), ErrorCode> {
```
Comment (alexandruradovici) [inline apis/ninedof/src/lib.rs]:
```suggestion
    pub fn read_gyro_sync() -> Result<NineDofData, ErrorCode> {
```

File: apis/ninedof/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+        listener: &'share NineDofListener<F>,
+        subscribe: Handle<Subscribe<'share, S, DRIVER_NUM, 0>>,
+    ) -> Result<(), ErrorCode> {
+        S::subscribe::<_, _, DefaultConfig, DRIVER_NUM, 0>(subscribe, listener)
+    }
+
+    /// Initiate a synchronous accelerometer measurement.
+    /// Returns Ok(accelerometer_value) if the operation was successful
+    /// Returns Err(ErrorCode) if the operation was unsuccessful
+    pub fn read_accelerometer_sync(x: &mut i32, y: &mut i32, z: &mut i32) -> Result<(), ErrorCode> {
```
Comment (alexandruradovici) [inline apis/ninedof/src/lib.rs]:
```suggestion
    pub fn read_accelerometer_sync() -> Result<NineDofData, ErrorCode> {
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: apis/ninedof/src/lib.rs (hunk 5 above)
Comment (SheepSeb) [reply apis/ninedof/src/lib.rs]: I have modified those and don't apper in the last commit, they're outdated

File: apis/ninedof/Cargo.toml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+name = "libtock_ninedof"
+version = "0.1.0"
+authors = [
+    "Tock Project Developers <tock-dev@googlegroups.com>",
+    "dcz <gihuac.dcz@porcupinefactory.org>",
+]
+license = "MIT/Apache-2.0"
+edition = "2018"
+repository = "https://www.github.com/tock/libtock-rs"
+description = "libtock ninedof driver"
```
Comment (alexandruradovici) [inline apis/ninedof/Cargo.toml:11]:
```suggestion
description = "libtock nine degrees of freedom driver"
```

File: apis/ninedof/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+                if let Ok(()) = Self::read_accelerometer() {
+                    while data_cell.get() == None {
+                        S::yield_wait();
+                    }
+                }
+            }
+        });
+
+        match data_cell.get() {
+            None => Err(ErrorCode::Busy),
```
Comment (alexandruradovici) [inline apis/ninedof/src/lib.rs:88]:
```suggestion
            None => Err(ErrorCode::Busy),
```

This is one of the return values of the first two system calls. Same for the next ones.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: apis/ninedof/src/lib.rs (hunk 7 above)
Comment (SheepSeb) [reply apis/ninedof/src/lib.rs:88]: I don't understand what should I return here, ```Err(ErrorCode::Busy)``` or what error should I use?

Comment (alexandruradovici) [reply apis/ninedof/src/lib.rs:88]: Return the error that `subscribe` or `read_accelerometer` return.

**Comment by alexandruradovici** [comment]
@SheepSeb any updates?

**Review by alexandruradovici — DISMISSED** [review]
Looks good, let's send it to upstream.

**Comment by SheepSeb** [comment]
Added to the upstream [tock#468](https://github.com/tock/libtock-rs/pull/468)

**Comment by SheepSeb** [comment]
Merged into master

**Pull request closed without merging**